Use forward-declarations in EditorPlugin where possible

trollodel opened the pull request:
Extracted from #58506.
Follow up of #57306.
Follow up of #58136.

Reduce the number of EditorPlugin includes as much as possible (3 for now) to reduce the risks of recompiling all the editor plugins (so half the editor) by doing an unrelated changed. It probably speeds up a bit the compile time.

Also add some forward declarations in touched headers.

KoBeWi commented:
I see some new includes in headers. I think you could get rid of them once #65062 is merged, because EditorUndoRedoManager has some include hell right now.

trollodel commented:
Do you mean UndoRedo includes? I can remove them in the next rebase.

KoBeWi commented:
I mean anything you added in `h` files. None of it is necessary if you are not inheriting the class and not using it in definitions.

E.g. this
https://github.com/godotengine/godot/pull/60684/files#diff-d5abd655f4f5a08af5c64775f79f7371ba7e8ffb32e5a42bec2e0bd9d3dec694R36
or this
https://github.com/godotengine/godot/pull/60684/files#diff-620b639db0b2a91194289810bb63014b2ee72cf007ce75542a5c9185a3fb6dc6R36
and here you even removed forward-declaration for some reason:
https://github.com/godotengine/godot/pull/60684/files#diff-9271693e3cad79625c66b204b7823aa72ad25ee5791feadcf8fe2e740dce9941L56

trollodel commented:
> E.g. this
[#60684 (files)](https://github.com/godotengine/godot/pull/60684/files#diff-d5abd655f4f5a08af5c64775f79f7371ba7e8ffb32e5a42bec2e0bd9d3dec694R36)
or this
[#60684 (files)](https://github.com/godotengine/godot/pull/60684/files#diff-620b639db0b2a91194289810bb63014b2ee72cf007ce75542a5c9185a3fb6dc6R36)

The files you linked aren't EditorPlugins, so I need to check if the changes are necessary at all.

> and here you even removed forward-declaration for some reason:
[#60684 (files)](https://github.com/godotengine/godot/pull/60684/files#diff-9271693e3cad79625c66b204b7823aa72ad25ee5791feadcf8fe2e740dce9941L56)

EditorUndoRedoManager needs to be included (it's done above), because IIRC it fails to compile on plugin registration if it's not included.

KoBeWi commented:
> EditorUndoRedoManager needs to be included (it's done above), because IIRC it fails to compile on plugin registration if it's not included.

That's what the PR I mentioned solves.

trollodel commented:
> > EditorUndoRedoManager needs to be included (it's done above), because IIRC it fails to compile on plugin registration if it's not included.
> 
> That's what the PR I mentioned solves.

Oh, I see. Yes, I can remove it once that PR is merged.

trollodel commented:
> > > EditorUndoRedoManager needs to be included (it's done above), because IIRC it fails to compile on plugin registration if it's not included.
> > 
> > 
> > That's what the PR I mentioned solves.
> 
> Oh, I see. Yes, I can remove it once that PR is merged.

@KoBeWi I fixed the comment about EditorUndoRedoManager. I also checked changes in non editor plugin files, and they are needed because they include indirectly `editor_plugin.h`.

KoBeWi commented on editor/editor_feature_profile.h:
```diff
@@ -33,6 +33,7 @@
 
 #include "core/io/file_access.h"
 #include "core/object/ref_counted.h"
+#include "editor/editor_file_dialog.h"
```
This can be replaced with forward declaration.

KoBeWi commented:
I still don't get why you added new includes. If something is required indirectly, it can still be cleaned up, until everything compiles.

trollodel commented:
Let me clarify, `editor_plugin.h` used to include indirectly lots of stuff. Here I removed most of them and replace its dependencies with forward declarations (as for PR title). Then I fixed every unsatisfied dependencies until it compiles. I used includes to be consistent with existing code. But as you said, every new include can be replaced with forward declarations and I can do it if wanted.

KoBeWi commented:
> But as you said, every new include can be replaced with forward declarations and I can do it if wanted.

Well, that's the purpose of this PR, no? We'll eventually want to do it everywhere, so it's ok if it goes a bit out of scope.

trollodel committed:
Use forward-declarations in EditorPlugin where possible

trollodel commented:
> Well, that's the purpose of this PR, no? We'll eventually want to do it everywhere, so it's ok if it goes a bit out of scope.

Not really. Anyway, I changed all the includes (where possible) with forward declarations and also reorder them in some places.

akien-mga commented:
Thanks!
